WordPress/openverse-catalog
# StockSnap popularity and popularity documentation

## Fixes


Fixes #182 by @zackkrida 

## Description


**Update:** StockSnap again modified their API for us to include 'raw' counts of downloads, views, and favorites. This PR now uses the downloads as the popularity metric for StockSnap.

- Add documentation for popularity
- Add stocksnap "favorites" field to popularity calculations

Sadly, this needs to wait, as there's a problem with StockSnap. StockSnap's api is focused on the presentation of this data, so numbers are actually automatically converted to strings like "1k" in the back-end, instead of on the client! Our popularity code expects numbers so sadly this fails. "favorites" is the only StockSnap popularity metric that is used so infrequently that it has all numbers < 1000, so it's not _yet_ an issue there. We'd have to ask Adrian at StockSnap to stop passing formatted numbers through the API and instead format on the client-side of StockSnap.

Still, this was a good exercise. 

## Technical details


The popularity code is pretty clever, and it was fun to learn how it works! Calculating the 85th percentile for each provider seems to find good results. I hope we can extend it to work with multiple metrics for each provider in the future.

## Tests


1. `just recreate`
2. Observe the presence of StockSnap in the `image_popularity_metrics` and `image_popularity_constants` DB tables
3. Visit airflow at http://localhost:9090
4. Enable the `tsv_to_postgres_loader` and `stocksnap_workflow` DAGs with the toggles.
5. Wait.
6. When stocksnap is done, and the tsvs have been added to the DB, enable the `recreate_image_popularity_calculation` workflow.
7. Wait.
8. When that's done, observe the calculated `val` and `constant` in the `image_popularity_constants` table for the StockSnap row.
9. Spot check a few of the images in the `image_view`, looking for values in the `standardized_popularity` column.

## Screenshots


## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
